bin/src/main.rs

```diff
@@ -468,6 +494,68 @@ async fn main() {
     }
 }
 
+async fn check_startup_lease_status(
+    start_option: Option<&str>,
+    claimed_lease_ref: &mut Option<String>,
+    client: Kube_Client,
+) {
+    if start_option == Some("on") {
+        // Attempt to claim lease N times, then move on.
+        info!("Getting agent-startup-lease (making limited attempts)");
+        let k8s_lease_api =
+            k8s::lease::get_k8s_lease_api(&std::env::var("NAMESPACE").unwrap(), client).await;
+        for i in 0..K8S_STARTUP_LEASE_RETRY_ATTEMPTS {
+            info!("Attempting connection: {}", i);
+            match get_available_lease(K8S_STARTUP_LEASE_LABEL, &k8s_lease_api).await {
+                Some(available_lease) => {
+                    info!("Lease available: {:?}", available_lease);
+                    k8s::lease::claim_lease(
+                        available_lease,
+                        std::env::var("POD_NAME").unwrap(),
+                        &k8s_lease_api,
+                        claimed_lease_ref,
+                    )
+                    .await;
+                    break;
+                }
+                None => {
+                    info!("No lease availabe at this time. Waiting 1 second...");
+                    tokio::time::sleep(Duration::from_millis(1000)).await;
+                }
+            };
+        }
+    } else if start_option == Some("always") {
+        // Keep trying to claim lease forever.
+        info!("Getting agent-startup-lease (trying forever)");
+        let k8s_lease_api =
+            k8s::lease::get_k8s_lease_api(&std::env::var("NAMESPACE").unwrap(), client).await;
+        loop {
+            match get_available_lease(K8S_STARTUP_LEASE_LABEL, &k8s_lease_api).await {
+                Some(available_lease) => {
+                    info!("Lease available: {:?}", available_lease);
+                    k8s::lease::claim_lease(
+                        available_lease,
+                        std::env::var("POD_NAME").unwrap(),
+                        &k8s_lease_api,
+                        claimed_lease_ref,
+                    )
+                    .await;
+                    break;
+                }
+                None => {
+                    info!("No lease availabe at this time. Waiting 1 second...");
+                    tokio::time::sleep(Duration::from_millis(1000)).await;
+                }
+            };
+        }
+    } else {
+        info!(
+            "Kubernetes cluster initialised, K8s startup lease set to: {:?}",
+            start_option,
+        )
+    }
```

Comment:
The code duplication isn't great. Can you use a while with a looping condition built from the value of the `start_option`? Incoming rust psuedocode

```rust
let max_attempts = match start_option {
  Some("on") => { info!("Getting agent-startup-lease (making limited attempts)"); K8S_STARTUP_LEASE_RETRY_ATTEMPTS },
  Some("always") => { info!("Getting agent-startup-lease (trying forever)"); -1 },
  _ =>{ info!("Initialised..."); return }
}

let attempts = 0
while (max_attempts == -1) || (attempts < max_attempts) {
  // the get lease code
}
```

Reply:
Thanks for the feedback @jakedipity. I've refactored base on your suggestion.